lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+use rdp::core::event::*;
+use rdp::model::error::*;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::net::TcpStream;
+use std::os::unix::io::AsRawFd;
+use std::ptr;
+use std::sync::{Arc, Mutex};
+
+struct Client {
+    rdp_client: RdpClient<TcpStream>,
+    tcp_fd: usize,
+}
+type SyncRdpClient = Arc<Mutex<Client>>;
+
+// Rust-side registry of clients, to allow Go to reference a specific client in calls after
+// connect_rdp.
+lazy_static! {
+    static ref RDP_CLIENTS: Arc<Mutex<HashMap<i64, SyncRdpClient>>> =
+        Arc::new(Mutex::new(HashMap::new()));
+}
+
+fn register_client(client_ref: i64, client: Client) {
+    RDP_CLIENTS
+        .lock()
+        .unwrap()
+        .insert(client_ref, Arc::new(Mutex::new(client)));
+}
+
+fn unregister_client(client_ref: &i64) {
+    RDP_CLIENTS.lock().unwrap().remove(client_ref);
+}
+
+#[derive(Debug)]
+enum ClientError {
+    RDP(Error),
+    ClientNotFound,
+    Other(String),
+}
+
+impl From<Error> for ClientError {
+    fn from(e: Error) -> ClientError {
+        ClientError::RDP(e)
+    }
+}
+
+fn with_client<F: FnMut(&SyncRdpClient) -> Result<(), ClientError>>(
+    client_ref: &i64,
+    mut f: F,
+) -> Result<(), ClientError> {
+    match RDP_CLIENTS.lock().unwrap().get(client_ref) {
+        Some(client) => f(client),
+        None => Err(ClientError::ClientNotFound),
+    }
+}
+
+// CGOString is a CGO-compatible definition of a UTF-8 string that can be used in Go and Rust.
+// This string is not null-terminated.
+#[repr(C)]
+pub struct CGOString {
+    // Memory is freed on the receiving side.
+    data: *mut u8,
+    len: u16,
+}
+
+// CGOError is just a CGOString used when returning error messages.
+// A "null" CGOError means no error, like in Go.
+type CGOError = CGOString;
+
+impl CGOString {
+    fn null() -> CGOString {
+        CGOString {
+            data: std::ptr::null_mut(),
+            len: 0,
+        }
+    }
+    fn is_null(&self) -> bool {
+        self.data == std::ptr::null_mut()
+    }
+}
+
+impl From<CGOString> for String {
+    fn from(s: CGOString) -> String {
+        unsafe { String::from_raw_parts(s.data, s.len.into(), s.len.into()) }
+    }
+}
+
+// Important: when converting String to CGOString, the caller is responsible for manually freeing
+// the memory at CGOString.data.
+impl From<String> for CGOString {
+    fn from(mut s: String) -> CGOString {
+        let cs = CGOString {
+            data: s.as_mut_ptr(),
+            len: s.len() as u16,
+        };
+        // Tell Rust to forget about the underlying memory, so that the caller can free it later.
+        std::mem::forget(s);
+        cs
+    }
+}
+
+// connect_rdp establishes an RDP connection to go_addr with the provided credentials and screen
+// size. If succeeded, the client is internally registered under client_ref. When done with the
+// connection, the caller must call close_rdp.
+#[no_mangle]
+pub extern "C" fn connect_rdp(
```

Comment:
This function should just be a thin wrapper to a more "rustic" function that utilizes `Result` for errors. This allows you to use the `?` operator for short circuiting error cases which is a lot nicer and foolproof. Generally extern c functions should only deal with conversions and calling another Rust-style function.